### ISSUE TITLE
Read-access to rootPage is a requirement to generate any preview

### ISSUE THREAD

**Issue by YKWeyer** [Yoast/Yoast-SEO-for-TYPO3#415]
* [x] I've read and understood the [contribution guidelines](https://github.com/Yoast/Yoast-SEO-for-TYPO3/blob/master/.github/CONTRIBUTING.md).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

### Please give us a description of what happened.

**TL;DR:** Users without read-access to the main page cannot generate previews for any page at all.

On our staging system, the TYPO3 Frontend is accessible only to logged-in editors with sufficient permissions. This behaviour is implemented using Typoscript conditions:
```
[globalVar = TSFE : beUserLogin < 1]
  page >
  page = PAGE
  page.10 = TEXT
  page.10.value = Backend login required!

  page.config {
    additionalHeaders {
      10 {
        header = Forbidden: Site access denied.
        replace = 0
        httpResponseCode = 403
      }

      20.header = Cache-control: Private
    }

    sendCacheHeaders = 1
    cache_period = 5
    no_cache = 1
  }

  # Restrict access to other typeNums all the same

  ## YOAST
  yoast_seo_sitemap < page
  yoast_seo_sitemap.typeNum = {$config.yoast_seo.sitemap_xml_type}

  yoast_seo_preview < page
  yoast_seo_preview.typeNum = {$config.yoast_seo.fe_preview_type}

[global]
```
This ensures editors cannot read pages outside of the tree they are allowed to access.

Yet, the preview URL (`/?type=1480321830&pageIdToCheck={pid}&languageIdToCheck={L}`) always uses the rootPage. If editors have no read permission on that specific page, they cannot generate any preview at all. 

One possible solution for us would be to remove the `yoast_seo_preview < page` section of the Typoscript conditions, but that would make it possible to bypass the permissions check simply by calling `?type=1480321830&pageIdToCheck={pid}&languageIdToCheck={L}` and virtually access the content of all the pages

### Please describe what you expected to happen and why.

Users should be able to generate previews for the pages they have access to, and shouldn't be able to generate previews for pages they have no access to.

It is not clear to me why the preview doesn't happen in the pages ID and language context. 
Something like `/?type=1480321830&pageIdToCheck={pid}&languageIdToCheck={L}&L={L}&id={pid}` would already indicate TYPO3 which page to render the preview for, and would use the native permission checks.

### How can we reproduce this behavior?

1. Add the test TypoScript to the rootpages Template
2. Create a "Test" page under the rootpage and add some content to it
3. When viewing the "Test" page module in backend, the preview should be displayed
4. Create an editor user with only access to a DB Mount pointing to the "Test" page
5. Impersonate this user
5. When viewing the "Test" page module in backend, the preview loads indefinitely. 
    * The developer console will display a 403 status for the URL /?type=1480321830&pageIdToCheck={pid}&languageIdToCheck=0

### Technical info
* TYPO3 version: 8.7.39
* Yoast SEO version: 7.2.0
* Relevant extensions in case of a bug:

**Comment by RinyVT** [comment]
Hi @YKWeyer,

Although I do get your point that we could also include the `id` parameter on the check, I do not think that we will adjust this.
This is because of a couple of reasons:
- We have just released 7.2.0, which would be the last release that supports CMS8. We will keep the 7.x branch under development for high-priority bugfixes but adding this parameter could potentially cause more issues than it fixes.
- From CMS9 and upwards, we make use of the Routing within core, which completely eliminates this problem as the full URL will be called, so no requests to the rootpage.
- I feel like this is an edge-case, it's not a very common configuration to deny access to the frontend this way

I think the easiest way to tackle this problem, is just to `XCLASS` our class `YoastSeoForTypo3\YoastSeo\Service\UrlService` and overwrite the `getPreviewUrl` as you said, with:

```
    public function getPreviewUrl(
        int $pageId,
        int $languageId,
        $additionalGetVars = ''
    ): string {
        $rootLine = BackendUtility::BEgetRootLine($pageId);
        // Mount point overlay: Set new target page id and mp parameter
        $pageRepository = GeneralUtility::makeInstance(PageRepository::class);
        $finalPageIdToShow = $pageId;
        $mountPointInformation = $pageRepository->getMountPointInfo($pageId);
        if ($mountPointInformation && $mountPointInformation['overlay']) {
            // New page id
            $finalPageIdToShow = $mountPointInformation['mount_pid'];
            $additionalGetVars .= '&MP=' . $mountPointInformation['MPvar'];
        }

        return $this->getUrlForType(self::FE_PREVIEW_TYPE, '&pageIdToCheck=' . $pageId . '&languageIdToCheck=' . $languageId . '&L=' . $languageId . '&id=' . $finalPageIdToShow);
    }
```

**Comment by YKWeyer** [comment]
Hi @RinyVT ,

I completely understand the strategical aspect of not supporting it, and I fully agree this is an edge-case. Thank you for suggesting a fix, this was also our fallback plan.

We implemented the XClass as you suggested, with the exception of a couple variable definitions that weren't used. This is our final version, should anyone else need it:

```php
    public function getPreviewUrl(
        int $pageId,
        int $languageId,
        $additionalGetVars = ''
    ): string {
        // Mount point overlay: Set new target page id and mp parameter
        $pageRepository = GeneralUtility::makeInstance(PageRepository::class);
        $finalPageIdToShow = $pageId;
        $mountPointInformation = $pageRepository->getMountPointInfo($pageId);
        if ($mountPointInformation && $mountPointInformation['overlay']) {
            // New page id
            $finalPageIdToShow = $mountPointInformation['mount_pid'];
        }

        return $this->getUrlForType(self::FE_PREVIEW_TYPE, '&pageIdToCheck=' . $pageId . '&languageIdToCheck=' . $languageId . '&L=' . $languageId . '&id=' . $finalPageIdToShow);
    }
```

Thanks again for your help!
